File: lib/models/magic_config.dart
import 'package:json_annotation/json_annotation.dart';

part 'magic_config.g.dart';

@JsonSerializable(explicitToJson: true)
class MagicConfig {
  final List<MagicStage> stages;
  final double blowThreshold; // Decibels, e.g. -10.0

  MagicConfig({
    required this.stages,
    this.blowThreshold = -10.0,
  });

  factory MagicConfig.fromJson(Map<String, dynamic> json) =>
      _$MagicConfigFromJson(json);
  Map<String, dynamic> toJson() => _$MagicConfigToJson(this);
}

enum StageType {
  @JsonValue('image')
  image,
  @JsonValue('video')
  video,
}

enum PlaybackMode {
  @JsonValue('one_shot')
  oneShot,
  @JsonValue('loop')
  loop,
}

enum TriggerType {
  @JsonValue('auto')
  auto, // Finishes and goes to next
  @JsonValue('tap_1')
  tap1, // Single tap
  @JsonValue('tap_2')
  tap2, // Double tap
  @JsonValue('tap_3')
  tap3, // Triple tap
  @JsonValue('long_press')
  longPress,
  @JsonValue('blow')
  blow,
  @JsonValue('none')
  none, // Waits for explicit jump or stays loop
}

@JsonSerializable()
class MagicStage {
  final String id;
  final StageType type;
  final String source; // filename in assets or local path
  final PlaybackMode mode;
  final List<MagicTrigger> triggers;

  // Helper for UI editor
  final String description;
  final bool showTime; // Show system time overlay

  MagicStage({
    required this.id,
    required this.type,
    required this.source,
    required this.mode,
    required this.triggers,
    this.description = '',
    this.showTime = false,
  });

  factory MagicStage.fromJson(Map<String, dynamic> json) =>
      _$MagicStageFromJson(json);
  Map<String, dynamic> toJson() => _$MagicStageToJson(this);
}

@JsonSerializable()
class MagicTrigger {
  final TriggerType type;
  final String nextStageId;
  final String? action; // 'exit', 'restart'

  MagicTrigger({
    required this.type,
    required this.nextStageId,
    this.action,
  });

  factory MagicTrigger.fromJson(Map<String, dynamic> json) =>
      _$MagicTriggerFromJson(json);
  Map<String, dynamic> toJson() => _$MagicTriggerToJson(this);
}

File: imagic_native/Sources/Models/MagicConfig.swift

import Foundation

// MARK: - Enums

enum StageType: String, Codable {
    case image
    case video
}

enum PlaybackMode: String, Codable {
    case oneShot = "one_shot"
    case loop
}

enum TriggerType: String, Codable, CaseIterable {
    case auto
    case tap1 = "tap_1"
    case tap2 = "tap_2"
    case tap3 = "tap_3"
    case longPress = "long_press"
    case blow
    case circle // New
    case none
}

// MARK: - Models

struct MagicConfig: Codable {
    var stages: [MagicStage]
    var blowThreshold: Double // e.g. -10.0
    
    init(stages: [MagicStage] = [], blowThreshold: Double = -10.0) {
        self.stages = stages
        self.blowThreshold = blowThreshold
    }
}

struct MagicStage: Codable, Identifiable {
    var id: String
    var type: StageType
    var source: String // filename or path
    var mode: PlaybackMode
    var triggers: [MagicTrigger]
    var description: String
    var showTime: Bool
    
    init(id: String = UUID().uuidString,
         type: StageType = .video,
         source: String = "",
         mode: PlaybackMode = .oneShot,
         triggers: [MagicTrigger] = [],
         description: String = "",
         showTime: Bool = false) {
        self.id = id
        self.type = type
        self.source = source
        self.mode = mode
        self.triggers = triggers
        self.description = description
        self.showTime = showTime
    }
}

struct MagicTrigger: Codable, Identifiable {
    var id: String = UUID().uuidString
    var type: TriggerType
    var nextStageId: String
    var action: String? // "exit", etc.
    
    enum CodingKeys: String, CodingKey {
        case type, nextStageId, action
    }
    
    init(type: TriggerType, nextStageId: String, action: String? = nil) {
        self.type = type
        self.nextStageId = nextStageId
        self.action = action
    }
}
